File: src/main/java/hu/frontrider/arcana/capabilities/CreatureEnchantProvider.java
package hu.frontrider.arcana.capabilities;

import net.minecraft.nbt.NBTBase;
import net.minecraft.util.EnumFacing;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.common.capabilities.CapabilityInject;
import net.minecraftforge.common.capabilities.ICapabilitySerializable;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

@SuppressWarnings("ALL")
public class CreatureEnchantProvider implements ICapabilitySerializable<NBTBase> {

    @CapabilityInject(ICreatureEnchant.class)
    public static final Capability<ICreatureEnchant> CREATURE_ENCHANT_CAPABILITY = null;

    private ICreatureEnchant instance = CREATURE_ENCHANT_CAPABILITY.getDefaultInstance();

    @Override
    public boolean hasCapability(@Nonnull Capability<?> capability, @Nullable EnumFacing facing) {
        return capability == CREATURE_ENCHANT_CAPABILITY;
    }

    @Nullable
    @Override
    public <T> T getCapability(@Nonnull Capability<T> capability, @Nullable EnumFacing facing) {
        return capability == CREATURE_ENCHANT_CAPABILITY ? CREATURE_ENCHANT_CAPABILITY.cast(this.instance) : null;
    }

    @Override
    public NBTBase serializeNBT() {
        return CREATURE_ENCHANT_CAPABILITY.getStorage().writeNBT(CREATURE_ENCHANT_CAPABILITY, this.instance, null);
    }

    @Override
    public void deserializeNBT(NBTBase nbt) {
        CREATURE_ENCHANT_CAPABILITY.getStorage().readNBT(CREATURE_ENCHANT_CAPABILITY, this.instance, null, nbt);
    }
}

File: src/main/java/hu/frontrider/arcana/capabilities/creatureenchant/CreatureEnchantProvider.kt
package hu.frontrider.arcana.capabilities.creatureenchant

import hu.frontrider.arcana.capabilities.inhibitor.IInhibitor
import net.minecraft.nbt.NBTBase
import net.minecraft.util.EnumFacing
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.common.capabilities.CapabilityInject
import net.minecraftforge.common.capabilities.ICapabilitySerializable

class CreatureEnchantProvider : ICapabilitySerializable<NBTBase> {

    private val instance = CREATURE_ENCHANT_CAPABILITY!!.defaultInstance

    override fun hasCapability(capability: Capability<*>, facing: EnumFacing?): Boolean {
        return capability === CREATURE_ENCHANT_CAPABILITY
    }

    override fun <T> getCapability(capability: Capability<T>, facing: EnumFacing?): T? {
        return if (capability === CREATURE_ENCHANT_CAPABILITY) CREATURE_ENCHANT_CAPABILITY.cast<T>(this.instance) else null
    }

    override fun serializeNBT(): NBTBase? {
        return CREATURE_ENCHANT_CAPABILITY.storage.writeNBT(CREATURE_ENCHANT_CAPABILITY, this.instance, null)
    }

    override fun deserializeNBT(nbt: NBTBase) {
        CREATURE_ENCHANT_CAPABILITY.storage.readNBT(CREATURE_ENCHANT_CAPABILITY, this.instance, null, nbt)
    }

    companion object {

        @CapabilityInject(ICreatureEnchant::class)
        lateinit var CREATURE_ENCHANT_CAPABILITY: Capability<ICreatureEnchant>
    }
}